File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/RestrictDeserializedTypes.LosFormatter.cs
using System;
using System.Web.UI;

internal class RestrictDeserializedTypes
{
    public void DefaultConstructor()
    {
        new LosFormatter();                                                 // FIXME Non-compliant {{Serialized data signature (MAC) should be verified.}}
        new LosFormatter { };                                               // FIXME Non-compliant
    }

    public void LiteralExpression()
    {
        new LosFormatter(false, "");                                        // FIXME Non-compliant {{Serialized data signature (MAC) should be verified.}}
        new LosFormatter(false, new byte[0]);                               // FIXME Non-compliant {{Serialized data signature (MAC) should be verified.}}

        new LosFormatter(true, "");                                         // Compliant - MAC filtering is enabled
        new LosFormatter(true, new byte[0]);                                // Compliant - MAC filtering is enabled

        new LosFormatter(macKeyModifier: new byte[0], enableMac: false);    // FIXME Non-compliant
        new LosFormatter(macKeyModifier: new byte[0], enableMac: true);     // Compliant
    }

    public void FunctionParameter(bool condition)
    {
        new LosFormatter(condition, "");                                    // Compliant - unsure about condition value

        if (condition)
        {
            new LosFormatter(condition, "");                                // Compliant - MAC filtering is enabled
        }
        else
        {
            new LosFormatter(condition, "");                                // FIXME Non-compliant - MAC filtering should be enabled
        }
    }

    public void LocalVariables()
    {
        var trueVar = true;
        new LosFormatter(trueVar, "");                                      // Compliant - MAC filtering is enabled

        var falseVar = false;
        new LosFormatter(falseVar, "");                                     // FIXME Non-compliant - MAC filtering should be enabled
    }

    public void TernaryOp(bool condition)
    {
        var falseVar = condition ? false : false;
        new LosFormatter(falseVar, "");                                     // FIXME Non-compliant - MAC filtering should be enabled

        var trueVar = condition ? true : true;
        new LosFormatter(trueVar, "");

        new LosFormatter(condition ? false : true, "");
    }

    public LosFormatter ExpressionBodyFalse() =>
        new LosFormatter(false, "");                                        // FIXME Non-compliant - MAC filtering should be enabled

    public LosFormatter ExpressionBodyTrue() =>
        new LosFormatter(true, "");                                         // Compliant - MAC filtering is enabled

    public void InLambdaFunction()
    {
        Func<LosFormatter> createSafe = () => new LosFormatter(true, "");   // Compliant - MAC filtering is enabled
        Func<LosFormatter> createUnsafe = () => new LosFormatter(true, ""); // Compliant - FN: lambda functions are not scanned by symbolic execution
    }

    public LosFormatter Switch(bool condition)
    {
        switch (condition)
        {
            case true:
                return new LosFormatter(condition, "");
            default:
                return new LosFormatter(condition, "");                     // Compliant - FN: lambda functions are not scanned by symbolic execution
        };
    }

    public void DataFlow()
    {
        var condition = false;
        condition = true;
        if (condition)
        {
            new LosFormatter(condition, "");
        }
        else
        {
            new LosFormatter(condition, "");                                // Unreachable
        }

        condition = false;
        if (condition)
        {
            new LosFormatter(condition, "");                                // Unreachable
        }
        else
        {
            new LosFormatter(condition, "");                                // FIXME Non-compliant - MAC filtering should be enabled
        }

        new LosFormatter(new bool(), "");
    }
}

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/RestrictDeserializedTypes.LosFormatter.vb
Imports System
Imports System.Web.UI

Friend Class RestrictDeserializedTypes
    Public Sub DefaultConstructor()
        New LosFormatter()
    End Sub

    Public Sub LiteralExpression()
        New LosFormatter(False, "")
        New LosFormatter(False, New Byte(-1) {})
        New LosFormatter(True, "")
        New LosFormatter(True, New Byte(-1) {})
        New LosFormatter(macKeyModifier:=New Byte(-1) {}, enableMac:=False)
        New LosFormatter(macKeyModifier:=New Byte(-1) {}, enableMac:=True)
    End Sub

    Public Sub FunctionParameter(ByVal condition As Boolean)
        New LosFormatter(condition, "")

        If condition Then
            New LosFormatter(condition, "")
        Else
            New LosFormatter(condition, "")
        End If
    End Sub

    Public Sub LocalVariables()
        Dim trueVar = True
        New LosFormatter(trueVar, "")
        Dim falseVar = False
        New LosFormatter(falseVar, "")
    End Sub

    Public Sub TernaryOp(ByVal condition As Boolean)
        Dim falseVar = If(condition, False, False)
        New LosFormatter(falseVar, "")
        Dim trueVar = If(condition, True, True)
        New LosFormatter(trueVar, "")
        New LosFormatter(If(condition, False, True), "")
    End Sub

    Public Function ExpressionBodyFalse() As LosFormatter
        Return New LosFormatter(False, "")
    End Function

    Public Function ExpressionBodyTrue() As LosFormatter
        Return New LosFormatter(True, "")
    End Function

    Public Sub InLambdaFunction()
        Dim createSafe As Func(Of LosFormatter) = Function() New LosFormatter(True, "")
        Dim createUnsafe As Func(Of LosFormatter) = Function() New LosFormatter(True, "")
    End Sub

    Public Function Switch(ByVal condition As Boolean) As LosFormatter
        Select Case condition
            Case True
                Return New LosFormatter(condition, "")
            Case Else
                Return New LosFormatter(condition, "")
        End Select
    End Function

    Public Sub DataFlow()
        Dim condition = False
        condition = True

        If condition Then
            New LosFormatter(condition, "")
        Else
            New LosFormatter(condition, "")
        End If

        condition = False

        If condition Then
            New LosFormatter(condition, "")
        Else
            New LosFormatter(condition, "")
        End If

        New LosFormatter(New Boolean(), "")
        New LosFormatter(Nothing, "")
    End Sub

End Class
